odoo/tools/populate.py

```diff
@@ -1,180 +1,357 @@
+        """
+        query = SQL("""
+        SELECT EXISTS(SELECT 1
+              FROM pg_index idx
+                   JOIN pg_class t ON t.oid = idx.indrelid
+                   JOIN pg_class i ON i.oid = idx.indexrelid
+                   JOIN pg_attribute a ON a.attnum = ANY (idx.indkey) AND a.attrelid = t.oid
+              WHERE t.relname = %s  -- tablename
+                AND a.attname = %s  -- column
+                AND idx.indisunique = TRUE) AS is_unique;
+        """, _model._table, _field.name)
+        return _env.execute_query(query)[0][0]
+
+    # many2one fields are not considered, as a name_search would resolve it to the _rec_names_search of the related model
+    if model._rec_names_search and field.name in model._rec_names_search and field.type != 'many2one':
+        return True
+    if field.type in ('date', 'datetime'):
+        return True
+    if field.index == 'trigram':
+        return True
+    return is_unique(env, model, field)
+
+
+def variate_field(env, model, field, table_alias, series_alias, factors):
+    """
+    Returns a variation of the source field,
+    to avoid unique constraint, or better distribute data.
+
+    :return: a str representing the source column, or an SQL(expression/subquery)
+    """
+    match field.type:
+        case 'char' | 'text':
+            return vary_char_field(env, model, field, postfix=series_alias)
+        case 'date' | 'datetime':
+            return vary_date_field(env, model, factors)
+        case 'html':
+            # For the sake of simplicity we don't vary html fields
+            return field.name
+        case _:
+            _logger.warning("The field %s of type %s was marked to be varied, "
+                            "but no variation branch was found! Defaulting to a raw copy.", field, field.type)
+            # fallback on a raw copy
+            return field.name
+
+
+def fetch_last_id(env, model):
+    query = SQL('SELECT id FROM %s ORDER BY id DESC LIMIT 1', SQL.identifier(model._table))
+    return env.execute_query(query)[0][0]
+
+
+def populate_field(env, model, field, populated, factors, table_alias='t', series_alias='s'):
+    """
+    Returns a tuple representing the destination and the source expression (str column or SQL expression)
+    """
+    def copy_noop():
+        return None, None
+
+    def copy_raw(_field):
+        return _field.name, _field.name
+
+    def copy(_field):
+        if field_need_variation(env, model, _field):
+            return _field.name, variate_field(env, model, _field, table_alias, series_alias, factors)
+        else:
+            return copy_raw(_field)
+
+    def copy_id():
+        last_id = fetch_last_id(env, model)
+        populated[model] = last_id  # this adds the model in the populated dict
+        return field.name, SQL('id + %(last_id)s * %(series_alias)s', last_id=last_id, series_alias=SQL.identifier(series_alias))
+
+    def copy_many2one(_field):
```

Comment:
You misuse standard Python conventions.  Prefixing a variable with `_` is meant to prevent a linter warning for some _unused_ variable.  The usual convention in this context would be `field_`.